### PR TITLE
Use errors instead of exceptions when creating a PFFile.

### LINKED ISSUE #338: Improve error messages when creating an invalid PFFile.
Currently, if you attempt to create a `PFFile` with a `nil` NSData parameter, you get a strange 'file not found' error when attempting to save it to Parse.

Let's make this either 

 - Throw an exception
 - Make the PFFile constructor return `nil`, when invalid data is passed in.

Somewhat relates to #6.

### PR DESCRIPTION
**Pull request by richardjrossiii**
Switching to errors provides a more consistent interface, as we should never use exceptions in objective-c except for unrecoverable errors.

Fixes #338.

### CODE REVIEW

File: Parse/PFFile.m [hunk 1]
```diff
@@ -73,12 +73,26 @@ + (instancetype)fileWithName:(NSString *)name contentsAtPath:(NSString *)path {
...
-    PFParameterAssert([fileManager fileExistsAtPath:path isDirectory:&directory] && !directory,
-                      @"%@ is not a valid file path for a PFFile.", path);
 
-    NSDictionary *attributess = [fileManager attributesOfItemAtPath:path error:nil];
-    unsigned long long length = [attributess[NSFileSize] unsignedLongValue];
-    PFParameterAssert(length <= PFFileMaxFileSize, @"PFFile cannot be larger than %lli bytes", PFFileMaxFileSize);
+    if (![fileManager fileExistsAtPath:path isDirectory:&directory] || directory) {
+        NSString *message = [NSString stringWithFormat:@"Failed to create PFFile at path '%@': "
+                                                        "file does not exist.", path];
+        *error = [NSError errorWithDomain:PFParseErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: 2 things here:
- let's use `PFErrorUtilities` to create an error, since it enforces PFParseErrorDomain, as well as logs to console by default.
- We should fine a code that fits this best from existing error codes, or avoid using `PFParseErrorDomain`, since we have a guarantee that we never use anything other than errors defined in PFConstants (`PFErrorCode`) enum.

File: Parse/PFFile.m [hunk 2]
```diff
@@ -73,12 +73,26 @@ + (instancetype)fileWithName:(NSString *)name contentsAtPath:(NSString *)path {
...
+                                 userInfo:@{ NSLocalizedDescriptionKey: message }];
+        return nil;
+    }
+
+    NSDictionary *attributes = [fileManager attributesOfItemAtPath:path error:nil];
+    unsigned long long length = [attributes[NSFileSize] unsignedLongValue];
+    if (length > PFFileMaxFileSize) {
+        NSString *message = [NSString stringWithFormat:@"Failed to create PFFile at path '%@': "
+                                                        "file is larger than %lluMB.", path, (PFFileMaxFileSize >> 20)];
+        *error = [NSError errorWithDomain:PFParseErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: Same as for line 80.

File: Parse/PFFile.m [hunk 3]
```diff
@@ -100,8 +114,22 @@ + (instancetype)fileWithName:(NSString *)name
                         data:(NSData *)data
                  contentType:(NSString *)contentType
                        error:(NSError **)error {
-    PFParameterAssert([data length] <= PFFileMaxFileSize,
-                      @"PFFile cannot be larger than %llu bytes", PFFileMaxFileSize);
+    if (!data) {
+        NSString *message = @"Cannot create a PFFile with nil data.";
+        *error = [NSError errorWithDomain:PFParseErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: Same as for line 80.

File: Parse/PFFile.m [hunk 4]
```diff
@@ -100,8 +114,22 @@ + (instancetype)fileWithName:(NSString *)name
...
+                                     code:EINVAL
+                                 userInfo:@{ NSLocalizedDescriptionKey: message }];
+        return nil;
+    }
+
+    if ([data length] > PFFileMaxFileSize) {
+        NSString *message = [NSString stringWithFormat:@"Failed to create PFFile with data: "
+                                                        "data is larger than %lluMB.", (PFFileMaxFileSize >> 20)];
+        *error = [NSError errorWithDomain:PFParseErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: Same as for line 80.

**Comment by nlutsenko** [comment]
Looks great, minus usage of errors.
Also, I guess it's worthwile adding a test for `nil` NSData?

File: Parse/PFFile.m [hunk 5]
```diff
@@ -100,8 +115,22 @@ + (instancetype)fileWithName:(NSString *)name
                         data:(NSData *)data
                  contentType:(NSString *)contentType
                        error:(NSError **)error {
-    PFParameterAssert([data length] <= PFFileMaxFileSize,
-                      @"PFFile cannot be larger than %llu bytes", PFFileMaxFileSize);
+    if (!data) {
+        NSString *message = @"Cannot create a PFFile with nil data.";
+        *error = [NSError errorWithDomain:NSPOSIXErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: Let's use `NSCocoaErrorDomain` with maybe `NSFileNoSuchFileError`?

File: Parse/PFFile.m [hunk 6]
```diff
@@ -100,8 +115,22 @@ + (instancetype)fileWithName:(NSString *)name
...
+                                     code:EINVAL
+                                 userInfo:@{ NSLocalizedDescriptionKey: message }];
+        return nil;
+    }
+
+    if ([data length] > PFFileMaxFileSize) {
+        NSString *message = [NSString stringWithFormat:@"Failed to create PFFile with data: "
+                                                        "data is larger than %lluMB.", (PFFileMaxFileSize >> 20)];
+        *error = [NSError errorWithDomain:NSPOSIXErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: Let's use `NSCocoaErrorDomain` with `NSFileReadTooLargeError` for code.

File: Parse/PFFile.m [hunk 7]
```diff
@@ -73,12 +74,26 @@ + (instancetype)fileWithName:(NSString *)name contentsAtPath:(NSString *)path {
...
-    PFParameterAssert([fileManager fileExistsAtPath:path isDirectory:&directory] && !directory,
-                      @"%@ is not a valid file path for a PFFile.", path);
 
-    NSDictionary *attributess = [fileManager attributesOfItemAtPath:path error:nil];
-    unsigned long long length = [attributess[NSFileSize] unsignedLongValue];
-    PFParameterAssert(length <= PFFileMaxFileSize, @"PFFile cannot be larger than %lli bytes", PFFileMaxFileSize);
+    if (![fileManager fileExistsAtPath:path isDirectory:&directory] || directory) {
+        NSString *message = [NSString stringWithFormat:@"Failed to create PFFile at path '%@': "
+                                                        "file does not exist.", path];
+        *error = [NSError errorWithDomain:NSPOSIXErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: `NSCocoaErrorDomain` with `NSFileNoSuchFileError` for code.

File: Parse/PFFile.m [hunk 8]
```diff
@@ -73,12 +74,26 @@ + (instancetype)fileWithName:(NSString *)name contentsAtPath:(NSString *)path {
...
+                                 userInfo:@{ NSLocalizedDescriptionKey: message }];
+        return nil;
+    }
+
+    NSDictionary *attributes = [fileManager attributesOfItemAtPath:path error:nil];
+    unsigned long long length = [attributes[NSFileSize] unsignedLongValue];
+    if (length > PFFileMaxFileSize) {
+        NSString *message = [NSString stringWithFormat:@"Failed to create PFFile at path '%@': "
+                                                        "file is larger than %lluMB.", path, (PFFileMaxFileSize >> 20)];
+        *error = [NSError errorWithDomain:NSPOSIXErrorDomain
```
Comment (nlutsenko) [inline Parse/PFFile.m]: Same as line 129.

**Comment by nlutsenko** [comment]
Be aware - Travis failed miserably on this one.

**Comment by nlutsenko** [comment]
Ship it!
